File: src/main/com/natpryce/snodge/JsonPath.java
package com.natpryce.snodge;

import com.google.common.base.Function;
import com.google.common.base.Functions;
import com.google.common.base.Joiner;
import com.google.common.base.Predicate;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.natpryce.snodge.internal.JsonFunctions;

import java.util.Arrays;
import java.util.Map;

import static java.util.Arrays.asList;

@SuppressWarnings("ChainOfInstanceofChecks")
public class JsonPath implements Function<JsonElement, JsonElement> {
    public static final JsonPath root = new JsonPath(new Object[0]);

    private final Object[] steps;

    private JsonPath(Object[] steps) {
        this.steps = steps;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        JsonPath that = (JsonPath) o;

        return Arrays.equals(this.steps, that.steps);

    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(this.steps);
    }

    @Override
    public String toString() {
        return pathBitsToString(steps, steps.length);
    }

    public static JsonPath of(Object... path) {
        return new JsonPath(path);
    }

    public int size() {
        return steps.length;
    }

    public Object at(int n) {
        return steps[(steps.length + n)% steps.length];
    }

    public JsonPath extend(Object... morePath) {
        Object[] newPath = new Object[this.steps.length + morePath.length];
        System.arraycopy(this.steps, 0, newPath, 0, this.steps.length);
        System.arraycopy(morePath, 0, newPath, this.steps.length, morePath.length);
        return new JsonPath(newPath);
    }

    @Override
    public JsonElement apply(JsonElement json) {
        JsonElement result = json;

        for (int i = 0; i < steps.length; i++) {
            result = applyPathElement(json, i, result);
        }

        return result;
    }

    private JsonElement applyPathElement(JsonElement root, int i, JsonElement parent) {
        Object pathBit = steps[i];

        if (pathBit instanceof String) {
            String memberName = (String) pathBit;
            JsonObject object = jsonObjectWithProperty(root, i, parent, memberName);
            return object.get(memberName);
        } else if (pathBit instanceof Integer) {
            int index = (Integer) pathBit;
            JsonArray array = jsonArrayWithIndex(root, i, parent, index);
            return array.get(index);
        } else {
            throw new IllegalArgumentException("unexpected path element: " + pathBitsToString(steps, i));
        }
    }

    public boolean endsWith(Object... suffix) {
        return steps.length >= suffix.length
            && asList(suffix).equals(asList(steps).subList(size()-suffix.length, size()));
    }

    public boolean startsWith(JsonPath prefix) {
        return size() >= prefix.size()
            && asList(steps).subList(0, prefix.size()).equals(asList(prefix.steps));
    }

    public JsonElement map(JsonElement json, Function<? super JsonElement, ? extends JsonElement> f) {
        return map(json, steps.length, f);
    }

    public DocumentMutation map(final Function<? super JsonElement, ? extends JsonElement> f) {
        return new DocumentMutation() {
            @Override
            public JsonElement apply(JsonElement json) {
                return map(json, f);
            }
        };
    }

    public JsonElement replace(JsonElement root, JsonElement replacement) {
        return map(root, Functions.constant(replacement));
    }

    private JsonElement replaceElement(JsonElement root, JsonElement parent, int i, JsonElement replacement) {
        Object pathBit = steps[i];

        if (pathBit instanceof String) {
            String memberName = (String) pathBit;
            JsonObject original = jsonObjectWithProperty(root, i, parent, memberName);

            JsonObject replaced = new JsonObject();
            for (Map.Entry<String, JsonElement> entry : original.entrySet()) {
                if (!entry.getKey().equals(memberName)) {
                    replaced.add(entry.getKey(), entry.getValue());
                }
            }
            replaced.add(memberName, replacement);

            return replaced;

        } else if (pathBit instanceof Integer) {
            int index = (Integer) pathBit;
            JsonArray original = jsonArrayWithIndex(root, i, parent, index);

            JsonArray replaced = new JsonArray();
            for (int j = 0; j < index; j++) {
                replaced.add(original.get(j));
            }
            replaced.add(replacement);
            for (int j = index + 1; j < original.size(); j++) {
                replaced.add(original.get(j));
            }

            return replaced;

        } else {
            throw new IllegalArgumentException("unexpected path element: " + pathBitsToString(steps, i));
        }
    }

    public DocumentMutation remove() {
        return new DocumentMutation() {
            @Override
            public JsonElement apply(JsonElement input) {
                return remove(input);
            }
        };
    }

    public JsonElement remove(final JsonElement root) {
        final int lastIndex = steps.length - 1;

        return map(root, lastIndex, new Function<JsonElement, JsonElement>() {
            @Override
            public JsonElement apply(JsonElement input) {
                return removeElement(root, lastIndex, input, steps[lastIndex]);
            }
        });
    }

    private JsonElement removeElement(JsonElement root, int i, JsonElement parent, Object pathBit) {
        if (pathBit instanceof String) {
            String memberName = (String) pathBit;
            JsonObject original = jsonObjectWithProperty(root, i, parent, memberName);
            return JsonFunctions.removeObjectProperty(original, memberName);

        } else if (pathBit instanceof Integer) {
            int index = (Integer) pathBit;
            JsonArray original = jsonArrayWithIndex(root, i, parent, index);
            return JsonFunctions.removeArrayElement(original, index);

        } else {
            throw new IllegalArgumentException("unexpected path element: " + pathBitsToString(steps, i));
        }
    }

    private JsonElement map(JsonElement json, int pathLength, Function<? super JsonElement, ? extends JsonElement> f) {
        JsonElement[] parents = new JsonElement[pathLength + 1];
        parents[0] = json;

        for (int i = 0; i < pathLength; i++) {
            parents[i + 1] = applyPathElement(json, i, parents[i]);
        }

        JsonElement replaced = f.apply(parents[pathLength]);

        for (int i = pathLength - 1; i >= 0; i--) {
            replaced = replaceElement(json, parents[i], i, replaced);
        }

        return replaced;
    }

    private JsonObject jsonObjectWithProperty(JsonElement root, int i, JsonElement parent, String memberName) {
        check(parent.isJsonObject(), "expected object", steps, i, root);
        JsonObject original = parent.getAsJsonObject();
        check(original.has(memberName), "no such member", steps, i, root);
        return original;
    }

    private JsonArray jsonArrayWithIndex(JsonElement root, int i, JsonElement parent, int index) {
        check(parent.isJsonArray(), "expected array", steps, i, root);
        JsonArray array = parent.getAsJsonArray();
        check(array.size() > index, "index out of bounds", steps, i, root);
        return array;
    }

    private static void check(boolean isOk, String what, Object[] pathBits, int badOne, JsonElement json) {
        if (!isOk) {
            throw new IllegalArgumentException(what + " at " + pathBitsToString(pathBits, badOne + 1) + " in " + json);
        }
    }

    private static String pathBitsToString(Object[] pathBits, int count) {
        return "/" + Joiner.on("/").join(Arrays.copyOf(pathBits, count));
    }

    public static class functions {
        public static Predicate<JsonPath> endsWith(final Object... suffix) {
            return new Predicate<JsonPath>() {
                @Override
                public boolean apply(JsonPath path) {
                    return path.endsWith(suffix);
                }

                @Override
                public String toString() {
                    return "endsWith(..." + JsonPath.of(suffix) + ")";
                }
            };
        }

        public static Predicate<JsonPath> startsWith(Object... prefix) {
            return startsWith(JsonPath.of(prefix));
        }

        public static Predicate<JsonPath> startsWith(final JsonPath prefix) {
            return new Predicate<JsonPath>() {
                @Override
                public boolean apply(JsonPath path) {
                    return path.startsWith(prefix);
                }

                @Override
                public String toString() {
                    return "startsWith(" + prefix + ")";
                }
            };
        }
    }
}

File: src/main/com/natpryce/snodge/JsonPath.kt
package com.natpryce.snodge

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.natpryce.snodge.internal.JsonFunctions
import java.util.function.Predicate

data class JsonPath(
    private val steps: List<Any>

) : (JsonElement) -> JsonElement {
    
    
    override fun toString(): String {
        return pathBitsToString(steps, steps.size)
    }
    
    fun size(): Int {
        return steps.size
    }
    
    val isRoot: Boolean
        get() = size() == 0
    
    fun at(n: Int) = steps[(steps.size + n) % steps.size]
    
    fun extend(vararg morePath: Any) = JsonPath(steps + morePath.toList())
    
    override fun invoke(json: JsonElement): JsonElement {
        var result = json
        
        for (i in steps.indices) {
            result = applyPathElement(json, i, result)
        }
        
        return result
    }
    
    private fun applyPathElement(root: JsonElement, i: Int, parent: JsonElement): JsonElement {
        val pathBit = steps[i]
        
        if (pathBit is String) {
            return jsonObjectWithProperty(root, i, parent, pathBit).get(pathBit)
        }
        else if (pathBit is Int) {
            return jsonArrayWithIndex(root, i, parent, pathBit).get(pathBit)
        }
        else {
            throw IllegalArgumentException("unexpected path element: " + pathBitsToString(steps, i))
        }
    }
    
    fun endsWith(vararg suffix: Any) =
        steps.size >= suffix.size
            && suffix.toList() == steps.subList(size() - suffix.size, size())
    
    fun startsWith(prefix: JsonPath): Boolean {
        return size() >= prefix.size() && steps.subList(0, prefix.size()) == prefix.steps
    }
    
    fun map(json: JsonElement, f: (JsonElement) -> JsonElement) =
        map(json, steps.size, f)
    
    fun map(f: (JsonElement) -> JsonElement) = object : DocumentMutation {
        override fun invoke(p1: JsonElement) = map(p1, f)
    }
    
    fun replace(root: JsonElement, replacement: JsonElement) =
        map(root) { replacement }
    
    private fun replaceElement(root: JsonElement, parent: JsonElement, i: Int, replacement: JsonElement): JsonElement {
        val pathBit = steps[i]
        
        if (pathBit is String) {
            val original = jsonObjectWithProperty(root, i, parent, pathBit)
            
            return replaceObjectPropertyValue(original, pathBit, replacement)
            
        }
        else if (pathBit is Int) {
            val original = jsonArrayWithIndex(root, i, parent, pathBit)
            
            return replaceArrayElement(original, pathBit, replacement)
            
        }
        else {
            throw IllegalArgumentException("unexpected path element: " + pathBitsToString(steps, i))
        }
    }
    
    private fun replaceObjectPropertyValue(original: JsonObject, memberName: String, replacement: JsonElement): JsonElement {
        val replaced = JsonObject()
        
        for ((key, value) in original.entrySet()) {
            if (key == memberName) {
                replaced.add(memberName, replacement)
            }
            else {
                replaced.add(key, value)
            }
        }
        
        return replaced
    }
    
    private fun replaceArrayElement(original: JsonArray, index: Int, replacement: JsonElement): JsonElement {
        val replaced = JsonArray()
        
        for (j in 0..index - 1) {
            replaced.add(original.get(j))
        }
        replaced.add(replacement)
        for (j in index + 1..original.size() - 1) {
            replaced.add(original.get(j))
        }
        
        return replaced
    }
    
    fun remove(): DocumentMutation {
        return object : DocumentMutation {
            override fun invoke(p1: JsonElement) = remove(p1)
        }
    }
    
    fun remove(root: JsonElement): JsonElement {
        val lastIndex = steps.size - 1
        
        return map(root, lastIndex, { input -> removeElement(root, lastIndex, input, steps[lastIndex]) })
    }
    
    private fun removeElement(root: JsonElement, i: Int, parent: JsonElement, pathBit: Any): JsonElement {
        if (pathBit is String) {
            val original = jsonObjectWithProperty(root, i, parent, pathBit)
            return JsonFunctions.removeObjectProperty(original, pathBit)
            
        }
        else if (pathBit is Int) {
            val original = jsonArrayWithIndex(root, i, parent, pathBit)
            return JsonFunctions.removeArrayElement(original, pathBit)
            
        }
        else {
            throw IllegalArgumentException("unexpected path element: " + pathBitsToString(steps, i))
        }
    }
    
    private fun map(json: JsonElement, pathLength: Int, f: (JsonElement) -> JsonElement): JsonElement {
        val parents = arrayOfNulls<JsonElement>(pathLength + 1)
        parents[0] = json
        
        for (i in 0..pathLength - 1) {
            parents[i + 1] = applyPathElement(json, i, parents[i]!!)
        }
        
        var replaced: JsonElement = f.invoke(parents[pathLength]!!)
        
        for (i in pathLength - 1 downTo 0) {
            replaced = replaceElement(json, parents[i]!!, i, replaced)
        }
        
        return replaced
    }
    
    private fun jsonObjectWithProperty(root: JsonElement, i: Int, parent: JsonElement, memberName: String): JsonObject {
        check(parent.isJsonObject, "expected object", steps, i, root)
        val original = parent.asJsonObject
        check(original.has(memberName), "no such member", steps, i, root)
        return original
    }
    
    private fun jsonArrayWithIndex(root: JsonElement, i: Int, parent: JsonElement, index: Int): JsonArray {
        check(parent.isJsonArray, "expected array", steps, i, root)
        val array = parent.asJsonArray
        check(array.size() > index, "index out of bounds", steps, i, root)
        return array
    }
    
    object functions {
        @JvmStatic
        fun endsWith(vararg suffix: Any): Predicate<JsonPath> {
            return object : Predicate<JsonPath> {
                override fun test(path: JsonPath): Boolean {
                    return path.endsWith(*suffix)
                }
                
                override fun toString(): String {
                    return "endsWith(..." + JsonPath.of(*suffix) + ")"
                }
            }
        }
    
        @JvmStatic
        fun startsWith(vararg prefix: Any): Predicate<JsonPath> {
            return startsWith(JsonPath.of(*prefix))
        }
    
        @JvmStatic
        fun startsWith(prefix: JsonPath): Predicate<JsonPath> {
            return object : Predicate<JsonPath> {
                override fun test(path: JsonPath): Boolean {
                    return path.startsWith(prefix)
                }
                
                override fun toString(): String {
                    return "startsWith($prefix)"
                }
            }
        }
    }
    
    companion object {
        @JvmField
        val root = JsonPath(emptyList())
        
        @JvmStatic
        fun of(vararg path: Any): JsonPath {
            return JsonPath(path.toList())
        }
        
        @JvmStatic
        private fun check(isOk: Boolean, what: String, pathBits: List<Any>, badOne: Int, json: JsonElement) {
            if (!isOk) {
                throw IllegalArgumentException(what + " at " + pathBitsToString(pathBits, badOne + 1) + " in " + json)
            }
        }
        
        @JvmStatic
        private fun pathBitsToString(pathBits: List<Any>, count: Int): String {
            return pathBits.subList(0, count)
                .map { it.toString() }
                .joinToString(prefix = "/", separator = "/", postfix = "")
        }
    }
}
